File: app/src/main/java/edu/gatech/donatrix/model/Location.java
package edu.gatech.donatrix.model;

import java.io.Serializable;
import edu.gatech.donatrix.dao.Database;
import android.content.Context;


public class Location implements Serializable {
    private int key;
    private String name;
    private String latitude;
    private String longitude;
    private String address;
    private String city;
    private String state;
    private String zip;
    private LocationType locationType;
    private String number;
    private String website;
    private ItemManager inventory;

    public Location(String name) {
        this.name = name;
    }

    public Location(String[] info) {
        this.key = Integer.parseInt(info[0]);
        this.setName(info[1]);
        this.setLatitude(info[2]);
        this.setLongitude(info[3]);
        this.setAddress(info[4]);
        this.setCity(info[5]);
        this.setState(info[6]);
        this.setZip(info[7]);
        this.setLocationType(LocationType.valueOf(info[8].replace(" ", "").toUpperCase()));
        this.setNumber(info[9]);
        this.setWebsite(info[10]);
    }

    public int getKey() {
        return key;
    }

    private void setName(String name) {
        this.name = name;
    }

    public String getName() {
        return name;
    }

    private void setLatitude(String latitude) {
        this.latitude = latitude;
    }

    public String getLatitude() {
        return latitude;
    }

    private void setLongitude(String longitude) {
        this.longitude = longitude;
    }

    public String getLongitude() {
        return longitude;
    }

    private void setAddress(String address) {
        this.address = address;
    }

    public String getAddress() {
        return address;
    }

    private void setCity(String city) {
        this.city = city;
    }

    public String getCity() {
        return city;
    }

    private void setState(String state) {
        this.state = state;
    }

    public String getState() {
        return state;
    }

    private void setZip(String zip) {
        this.zip = zip;
    }

    public String getZip() {
        return zip;
    }

    private void setLocationType(LocationType locationType) {
        this.locationType = locationType;
    }

    public LocationType getLocationType() {
        return locationType;
    }

    private void setNumber(String number) {
        this.number = number;
    }

    public String getNumber() {
        return number;
    }

    private void setWebsite(String website) {
        this.website = website;
    }

    public String getWebsite() {
        return website;
    }

    public void addItem(Item item, Context context, LocationEmployee employee) {
        Database.getInstance(context).addItem(item, employee);
    }

    public void removeItem(Item item) {
        this.inventory.removeItem(item);
    }
    @Override
    public String toString() {
        return String.format("%s | %s | %s | %s | %s | %s | %s | %s | %s | %s", name, latitude, longitude, address, city, state, zip, locationType.getType(), number, website);
    }

    @Override
    public boolean equals(Object obj) {
        if (!(obj instanceof Location)) {
            return false;
        } else {
            return this.key == ((Location) obj).key && this.address.equals(((Location) obj).address) && this.latitude.equals(((Location) obj).latitude) && this.city.equals(((Location) obj).city) && this.locationType == ((Location) obj).locationType && this.name.equals(((Location) obj).name) && this.longitude.equals(((Location) obj).longitude) && this.number.equals(((Location) obj).number) && this.state.equals(((Location) obj).state) && this.website.equals(((Location) obj).website) && this.zip.equals(((Location) obj).zip);
        }
    }

    @Override
    public int hashCode() {
        return this.key * this.name.length();
    }
}

File: app/src/main/java/edu/gatech/donatrix/model/Location.kt
package edu.gatech.donatrix.model

import java.io.Serializable
import edu.gatech.donatrix.dao.Database
import android.content.Context


class Location : Serializable {
    var key: Int = 0 //Paul rename to id or uid & Davidson initialized it to 0 for Kotlin support
    var name: String? = null
        private set //Paul: rename to title
    var latitude: String? = null
        private set //Paul: extract into coordinate object
    var longitude: String? = null
        private set //Paul: extract into coordinate object
    var address: String? = null
        private set //Paul: extract into address object
    var city: String? = null
        private set //Paul: extract into address object
    var state: String? = null
        private set //Paul: extract into address object
    var zip: String? = null
        private set //Paul: extract into address object
    var locationType: LocationType? = null
        private set
    var number: String? = null
        private set //Paul: rename to phoneNumber or telephone
    var website: String? = null
        private set
    private val inventory: ItemManager? = null

    constructor(name: String) {
        this.name = name
    }

    //Paul: info needs to be of type HashMap<String><String>
    constructor(info: Array<String>) {
        this.key = Integer.parseInt(info[0])
        this.name = info[1]
        this.latitude = info[2]
        this.longitude = info[3]
        this.address = info[4]
        this.city = info[5]
        this.state = info[6]
        this.zip = info[7]
        //Paul: the line below violates LOD, changes in LocationType can break this file:
        this.locationType = LocationType.valueOf(info[8].replace(" ", "").toUpperCase())
        this.number = info[9]
        this.website = info[10]
    }

    fun addItem(item: Item, context: Context, employee: LocationEmployee) {
        Database.getInstance(context)!!.addItem(item, employee)
    }

    fun removeItem(item: Item) {
        this.inventory!!.removeItem(item)
    }

    override fun toString(): String {
        //Paul: ugly syntax
        return String.format("%s | %s | %s | %s | %s | %s | %s | %s | %s | %s", name, latitude, longitude, address, city, state, zip, locationType!!.type, number, website)
    }

    override fun equals(obj: Any?): Boolean {
        return if (obj !is Location) {
            false
        } else {
            //Paul: ugly syntax
            this.key == obj.key && this.address == obj.address && this.latitude == obj.latitude && this.city == obj.city && this.locationType == obj.locationType && this.name == obj.name && this.longitude == obj.longitude && this.number == obj.number && this.state == obj.state && this.website == obj.website && this.zip == obj.zip
        }
    }

    override//Paul: not a proper hashcode method. Low priority
    fun hashCode(): Int {
        return this.key * this.name!!.length
    }
}
